### PR TITLE
Add Same Site Cookie Support

### PR DESCRIPTION
**Pull request by LukasReschke**
This adds Same Site Cookie Support, a new security feature supported by Chrome 51 (to be released in May to the stable channel) and Firefox has intentions to add it in Q2 2016 as well.

SameSite Cookies allows to specify from which origins a cookie is sent. The policy of "lax" prevents the cookie from being sent to ownCloud from any other domain than the one that ownCloud is running on. Effectively this prevents CSRF attacks on any HTTP method other than GET which is a very important and big security hardening.

This implementation behaves as following:
1. For browsers: If the cookie is not sent AND cookies are supported then add the cookie and reload the page.
2. For DAV clients: If the cookie is not sent AND cookies are supported then add the cookie and issue a 503 status code.

The implementation has the advantage that it is not intrusive and does not break existing clients or pages.

**Todo:**
- [x] Add strict cookie requirement for all other endpoints than "index.php"
- [x] Add annotation to AppFramework to opt-in controllers to strict cookie requirement
- [x] Add strict cookie requirement to CSRF check 
- [x] Add unit tests 

<hr/>

http://www.sjoerdlangkemper.nl/2016/04/14/preventing-csrf-with-samesite-cookie-attribute/ has a nice overview and introduction to the topic.

### CODE REVIEW

File: apps/files/controller/apicontroller.php [hunk 1]
```diff
@@ -86,6 +86,7 @@ public function __construct($appName,
 	 *
 	 * @NoAdminRequired
 	 * @NoCSRFRequired
+	 * @StrictCookieRequired
```
Comment (rullzer) [inline apps/files/controller/apicontroller.php]: So this is opt-in? And not opt-out?

**Comment by rullzer** [comment]
Firefox still works over here :)

File: apps/files/controller/apicontroller.php (hunk 1 above)
Comment (LukasReschke) [inline apps/files/controller/apicontroller.php]: It's a little bit more complex. So we do have two cookies:

- a "lax" cookie that will be sent by external domains for GET requests. but nothing else This one is always required.
- a "strict" one, this is never sent by external domains.

So the problem with the strict one is: If somebody links to you a ownCloud page with that setting it won't open. It only will open if you come from the same origin. - Is is thus only enabled for:

- Routes that are CSRF protected (Origin is ownCloud)
- WebDAV
- OCS

So basically everywhere that you don't open directly. However, there are also cases where a regular controller warrants this type of protection. That's what this annotation does. (so basically Non-CSRF protected GET APIs that should get some hardening)

Comment (rullzer) [inline apps/files/controller/apicontroller.php]: Awesome thanks for the explination I missed that part.
Very cool stuff!

**Comment by georgehrke** [comment]
Please rebase

**Comment by MorrisJobke** [comment]
> Please rebase

@LukasReschke There are conflicts in here.

**Comment by rullzer** [comment]
Ah yes because of all the PSR-4 stuff.
I rebased. Git was very smart about it.

**Comment by mwaclawek** [comment]
@LukasReschke I've added `SameSite` support to my [cookie library](https://github.com/delight-im/PHP-Cookie) but it seems you've set this using a hard-coded `Set-Cookie` header already. May I ask why you're sending cookies for both policies and checking for both?
